core/src/test/java/org/elasticsearch/cluster/routing/allocation/BalancedSingleShardTests.java

```diff
@@ -0,0 +1,235 @@
+        assertEquals(Type.YES, rebalanceDecision.getCanRebalanceDecision().type());
+        assertEquals(Type.NO, rebalanceDecision.getFinalDecisionType());
+        assertNotNull(rebalanceDecision.getFinalExplanation());
+        assertEquals(clusterState.nodes().getSize() - 1, rebalanceDecision.getNodeDecisions().size());
+        assertNull(rebalanceDecision.getAssignedNodeId());
+    }
+
+    public void testSingleShardBalanceProducesSameResultsAsBalanceStep() {
+        AllocationDecider allocationDecider = new AllocationDecider(Settings.EMPTY) {
+            @Override
+            public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+                return Decision.YES;
+            }
+        };
+        AllocationDecider rebalanceDecider = new AllocationDecider(Settings.EMPTY) {
+            @Override
+            public Decision canRebalance(ShardRouting shardRouting, RoutingAllocation allocation) {
+                return Decision.YES;
+            }
+        };
+        List<AllocationDecider> allocationDeciders = Arrays.asList(rebalanceDecider, allocationDecider);
+        final String[] indices = { "idx1", "idx2", "idx3" };
+        BalancedShardsAllocator allocator = new BalancedShardsAllocator(Settings.EMPTY);
+        // create a cluster state with 3 indices, each with 1 started primary shard, and only
+        // one node initially so all 3 primary shards get allocated to the same shard
+        ClusterState clusterState = ClusterStateCreationUtils.state(1, indices, 1);
+        // add a new node so one of the primaries can be rebalanced there
+        DiscoveryNodes.Builder nodesBuilder = DiscoveryNodes.builder(clusterState.nodes());
+        nodesBuilder.add(newNode(randomAsciiOfLength(7)));
+        clusterState = ClusterState.builder(clusterState).nodes(nodesBuilder).build();
+        RoutingAllocation routingAllocation = newRoutingAllocation(
+            new AllocationDeciders(Settings.EMPTY, allocationDeciders), clusterState);
+        // allocate and get the node that is now relocating
+        allocator.allocate(routingAllocation);
+        ShardRouting shardToRebalance = null;
+        for (RoutingNode routingNode : routingAllocation.routingNodes()) {
+            List<ShardRouting> relocatingShards = routingNode.shardsWithState(ShardRoutingState.RELOCATING);
+            if (relocatingShards.size() == 1) {
+                shardToRebalance = relocatingShards.get(0);
+                break;
+            }
+        }
+
+        routingAllocation = newRoutingAllocation(new AllocationDeciders(Settings.EMPTY, allocationDeciders), clusterState);
+        ShardRouting shard = clusterState.getRoutingNodes().activePrimary(shardToRebalance.shardId());
+        RebalanceDecision rebalanceDecision = allocator.decideRebalance(shard, routingAllocation);
+        assertEquals(shardToRebalance.relocatingNodeId(), rebalanceDecision.getAssignedNodeId());
```

Comment:
Since this adds a single node, won't the decision for where to rebalance the shard always be the same? Should the test have multiple nodes so it can ensure it's not just because it's the only node that it can go to?